sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,553 @@
+hierarchy, since the ability of work-items to make progress depends on a
+combination of the forward progress guarantees provided by the work-item
+threads of execution and the forward progress guarantees provided by the
+sub-group, work-group and root-group threads of execution. Often, correct
+execution of a kernel requires knowledge of the forward progress guarantees
+of a certain subset of these threads of execution.
+
+To facilitate reasoning about work-item coordination, the forward progress
+guarantees of threads of execution in SYCL can be qualified by a _coordination
+scope_. This extension defines the following coordination scopes:
+
+- All work-item threads of execution spawned directly or indirectly from the
+ND-range kernel's initial root thread of execution are considered part of the
+same root-group scope.
+- All work-item threads of execution spawned directly or indirectly from the
+same work-group thread of execution are considered part of the same work-group
+scope.
+- All work-item threads of execution spawned from the same sub-group thread of
+execution are considered part of the same sub-group scope.
+- Each individual work-item thread of execution is the sole member of its own
+work-item scope.
+
+[NOTE]
+====
+Qualifying a thread of execution's forward progress guarantees with a coordination
+scope can be thought of as a shorthand for combining the forward progress
+guarantees provided by each thread of execution in the hierarchy.
+For example, consider a hypothetical implementation that creates a separate
+`std::thread` (providing concurrent forward progress guarantees) to execute
+each sub-group, but which permits an unbounded number of work-groups executed
+as tasks (providing parallel forward progress guarantees). With work-group
+coordination scope, each sub-group provides concurrent forward progress
+guarantees: this reflects the guarantees associated with each `std::thread`
+upon creation. With root-group coordination scope, each sub-group provides
+only parallel forward progress guarantees: each sub-group's progress guarantees
+are weakened to reflect that the creation of a sub-group depends upon a
+specific work-group (task) being scheduled and making progress.
+====
+
+
+=== Forward progress guarantees
+
+The `forward_progress_guarantee` `enum` is used to represent the three classes
+of forward progress guarantee defined in the {cpp} core language specification.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class forward_progress_guarantee {
+  concurrent,
+  parallel,
+  weakly_parallel
+};
+
+}
+----
+
+[NOTE]
+====
+This `enum` is defined similarly to the one defined in
+https://wg21.link/p2300[P2300: `std::execution`].
+====
+
+
+=== Execution scopes
+
+The `execution_scope` `enum` is used to represent the hierarchy of threads of
+execution within a SYCL implementation.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class execution_scope {
```

Comment:
I don't think we should view this as a fixed hierarchy.  There may be implementations that want to add additional execution scopes in the middle of the existing hierarchy.  For example, if it's possible to provide stronger forward progress guarantees for all work-groups executing on the same cluster of SMs on an NVIDIA GPU, it would make sense to add something like a  `cluster` execution scope between `work_group` and `root_group`.

Aside: These sorts of concerns are why we split things into a coordination and target scope.  Initially, we considered implicitly specifying a forward progress guarantee relative to the immediate parent (i.e. work-item forward progress guarantees would always be relative to the sub-group, and so on), but this would break user code unexpectedly if/when new layers were added to the hierarchy.  By making both scopes explicit, we can add new layers to the hierarchy safely.

Reply:
Yeah that's a good point, I agree it makes sense then to leave these without set values.